### PR TITLE
[Hotfix] Update the block write interface

### PR DESCRIPTION
**Pull request by zivkovicmilos**
# Description

This PR fixes the blockchain write interface to be more precise.

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [x] I have added sufficient documentation both in code, as well as in the READMEs

### CODE REVIEW

**Review by 0xAleksaOpacic — APPROVED** [review]
Looks great. Thank you for refactoring you have done!

File: protocol/syncer.go [hunk 1]
```diff
@@ -242,10 +242,11 @@ func (pw *progressionWrapper) runUpdateLoop(subscription blockchain.Subscription
 
 // stopProgression stops the progression tracking
 func (pw *progressionWrapper) stopProgression() {
+	pw.stopCh <- struct{}{}
```
Comment (lazartravica) [inline protocol/syncer.go:245]: Why did you move the message sending to the top of the function?

File: e2e/pos_test.go [hunk 2]
```diff
@@ -613,3 +615,171 @@ func TestPoS_StakeUnstakeWithinSameBlock(t *testing.T) {
...
+) (*ibftOp.Snapshot, error) {
+	snapshot, snapshotErr := client.GetSnapshot(ctx, &ibftOp.SnapshotReq{
+		Latest: false,
+		Number: blockNum,
+	})
+
+	return snapshot, snapshotErr
+}
+
+func getNearestNextEpochBlock(blockNum uint64, epochSize uint64) uint64 {
```
Comment (lazartravica) [inline e2e/pos_test.go]: can be renamed to simply `getNextEpochBlock`

File: e2e/pos_test.go [hunk 3]
```diff
@@ -613,3 +615,171 @@ func TestPoS_StakeUnstakeWithinSameBlock(t *testing.T) {
...
+	if epochSize > blockNum {
+		return epochSize
+	}
+
+	nearest := blockNum + epochSize/2
+	nearest = nearest - (nearest % epochSize)
+
+	if nearest < blockNum {
+		nearest += epochSize
+	}
```
Comment (lazartravica) [inline e2e/pos_test.go]: i feel we can make this more readable.

do you think this is faster than performing mod block size to check if we're already there, and then doing integer division and multiplication by epoch size to deduce the next epoch block?

File: e2e/pos_test.go [hunk 4]
```diff
@@ -613,3 +615,171 @@ func TestPoS_StakeUnstakeWithinSameBlock(t *testing.T) {
...
+
+	return nearest
+}
+
+func TestSnapshotUpdating(t *testing.T) {
+	fountainKey, fountainAddr := tests.GenerateKeyAndAddr(t)
```
Comment (lazartravica) [inline e2e/pos_test.go]: you mean faucet?

File: blockchain/blockchain.go [hunk 5]
```diff
@@ -519,43 +519,29 @@ func (b *Blockchain) WriteHeadersWithBodies(headers []*types.Header) error {
...
-	parent, ok := b.readHeader(blocks[0].ParentHash())
+	b.logger.Info(
+		"write block",
+		"num",
+		block.Number(),
+		"parent",
+		block.ParentHash(),
+	)
+
+	parent, ok := b.readHeader(block.ParentHash())
```
Comment (lazartravica) [inline blockchain/blockchain.go:538]: this is incurring a performance cost when syncing blocks. in the previous implementation we only checked if the first block's parent is present in the db. now we will check it all the time. 

can you benchmark the syncing tests with and without this line here just to see how much of a performance cost can we expect? 

do you see any simple and clean way we can skip this check when bulk syncing subsequent  blocks?

File: blockchain/blockchain.go [hunk 6]
```diff
@@ -564,94 +550,89 @@ func (b *Blockchain) WriteBlocks(blocks []*types.Block) error {
...
+		)
+	}
 
-			return fmt.Errorf(
-				"transaction root hash mismatch: have %s, want %s",
-				hash,
-				block.Header.TxRoot,
-			)
-		}
+	// TODO, the wrapper around transactions
```
Comment (lazartravica) [inline blockchain/blockchain.go]: i am not sure what this TODO means, do you know?

File: blockchain/blockchain.go [hunk 7]
```diff
@@ -564,94 +550,89 @@ func (b *Blockchain) WriteBlocks(blocks []*types.Block) error {
...
+		"hash", header.Hash,
+		"txns", len(block.Transactions),
+	}
+	if prevHeader, ok := b.GetHeaderByNumber(header.Number - 1); ok {
+		diff := header.Timestamp - prevHeader.Timestamp
+		logArgs = append(logArgs, "generation_time_in_sec", diff)
 	}
 
+	b.logger.Info("new block", logArgs...)
 	b.logger.Info("new head", "hash", b.Header().Hash, "number", b.Header().Number)
```
Comment (lazartravica) [inline blockchain/blockchain.go]: i don't think we need this `new head` message any more, since the previous `new block` will log all the relevant information.

**Review by lazartravica — CHANGES_REQUESTED** [review]
Good job on the tests 💪  Take a look at my review comments.

File: consensus/ibft/ibft_test.go [hunk 8]
```diff
@@ -537,7 +537,7 @@ func TestWriteTransactions(t *testing.T) {
 				{Nonce: 4, Gas: 10001}, // exceeds block gas limit
 				{Nonce: 5},             // included
 				{Nonce: 6},             // reaches gas limit - returned to pool
-				{Nonce: 7}},            // not considered - stays in pool
+				{Nonce: 7}}, // not considered - stays in pool
```
Comment (lazartravica) [inline consensus/ibft/ibft_test.go]: Format

File: protocol/syncer.go (hunk 1 above)
Comment (zivkovicmilos) [reply protocol/syncer.go:245]: `stopProgression` shouldn't be holding a lock at the moment it's sending a stop signal to the stop channel the local runner is listening on - since it can cause a deadlock situation where the thread is holding a lock and waiting for someone to read from the `stopCh` and some other thread wants to grab the lock

File: e2e/pos_test.go (hunk 2 above)
Comment (zivkovicmilos) [reply e2e/pos_test.go]: Fixed 👍 

d2cc7ff

File: e2e/pos_test.go (hunk 4 above)
Comment (zivkovicmilos) [reply e2e/pos_test.go]: Yep 👍 

Fixed 👍 

c831ab2

File: blockchain/blockchain.go (hunk 7 above)
Comment (zivkovicmilos) [reply blockchain/blockchain.go]: Agreed 💯 

Fixed 👍 

a90c517

File: consensus/ibft/ibft_test.go (hunk 8 above)
Comment (zivkovicmilos) [reply consensus/ibft/ibft_test.go]: Fixed 👍 

55ec67f

File: blockchain/blockchain.go (hunk 6 above)
Comment (zivkovicmilos) [reply blockchain/blockchain.go]: Honestly I have no idea, this comment was made ~3 years ago.

I've removed it since I can't see the relevance any more:

39e4783

File: blockchain/blockchain.go (hunk 5 above)
Comment (zivkovicmilos) [reply blockchain/blockchain.go:538]: You're completely right to assume it adds overhead - but that's something that we can only see when bulk syncing. Every other part of the PSDK already did this call on a per-block basis as they write 1 block at a time.

I don't see an easy way of skipping this check with the `WriteBlock` functionality, as it is a critical chain sanity check, it needs to be there at one point before pushing a block to the chain.

The header will be initially fetched from the in-memory cache and if there's a cache miss, it will grab it from LevelDB, either way it's a quick lookup as it's fetching only the header and not the entire block

File: e2e/pos_test.go (hunk 3 above)
Comment (zivkovicmilos) [reply e2e/pos_test.go]: Simplified it by a lot 👍 


8a2ee9c

File: protocol/syncer.go (hunk 1 above)
Comment (lazartravica) [reply protocol/syncer.go:245]: got it! thanks for the explanation 👍

**Review by lazartravica — APPROVED** [review]
🥳
